### PR TITLE
Capture output from threads spawned in tests

### LINKED ISSUE #42474: `cargo test` doesn't capture print from threads
I've filed this bug in [rust-lang/cargo](https://github.com/rust-lang/cargo/issues/4129) but I'm afraid rust-lang/rust is the correct repo for this.

I ran the tests for one of my crates that uses threads in its example code. I've noticed it printed some lines to stdout, even though `--nocapture` wasn't set, but not all of them. It appears that text printed from new threads are not correctly captured, I think this is not intended.

## Reproduce
```
#[cfg(test)]
mod tests {
    use std::thread;
    use std::time::Duration;

    #[test]
    fn it_works() {
        thread::spawn(move || {
            println!("request: {:?}", 123);
        });

        thread::sleep(Duration::from_secs(1));
        println!("response: {:?}", 456);
    }
```

## Output
```
running 2 tests
request: 123
test tests::it_works ... ok
[...]
```

## Meta

`rustc --version --verbose`:

```
rustc 1.17.0
binary: rustc
commit-hash: unknown
commit-date: unknown
host: x86_64-unknown-linux-gnu
release: 1.17.0
LLVM version: 4.0
```

### PR DESCRIPTION
**Pull request by SergioBenitez**
This is revival of #75172.

Original text:
> Fixes #42474.
> 
> r? @​dtolnay since you expressed interest in this, but feel free to redirect if you aren't the right person anymore.

---

Closes #75172.

### CODE REVIEW

File: library/test/src/helpers/sink.rs [hunk 1]
```diff
@@ -14,6 +15,12 @@ impl Sink {
     }
 }
 
+impl io::LocalOutput for Sink {
+    fn clone_box(&self) -> Box<dyn io::LocalOutput> {
+        Box::new(Self(self.0.clone()))
```
Comment (m-ou-se) [inline library/test/src/helpers/sink.rs:20]:
```suggestion
        Box::new(self.clone())
```

**Comment by m-ou-se** [comment]
Looks good. Thanks.

Did you look at this failure from the original PR? https://github.com/rust-lang/rust/pull/75172#issuecomment-675881357 Would that still be a problem?

**Comment by m-ou-se** [comment]
I think that failure is caused by the `LOCAL_STDERR` being destroyed before `HANDLE` is dropped in the spawned thread in that test:

https://github.com/rust-lang/rust/blob/6b9fbf212a06944ff23325d2d63215805dc3c6c3/src/test/ui/issues/issue-24313.rs#L16

Before this change, that thread would not have `LOCAL_STDERR` set. But now that it is set, the `panic!()` in the `Drop` of this test's `Handle` will cause the panic handler to access it after it was set (in thread::spawn) and already destroyed, triggering this panic:

https://github.com/rust-lang/rust/blob/6b9fbf212a06944ff23325d2d63215805dc3c6c3/library/std/src/thread/local.rs#L248-L251

**Comment by m-ou-se** [comment]
Looks like that test doesn't run in the test framework, so now that this change checks LOCAL_STREAMS before setting any streams, the failure should be fixed.

**Review by m-ou-se — APPROVED** [review]
Lgtm, with one tiny nit.

**Comment by m-ou-se** [comment]
Panicking in the Drop implementaiton of a thread local in a thread in a unit test is probably not very common, but that situation could be partially fixed by making the panic handler use `LOCAL_STDERR.try_with` instead of `.with`. (It wouldn't capture the panic, but at least it would no longer trigger a double-panic.) Might be good as a follow-up PR (including a test that that works).

**Comment by jyn514** [comment]
(FYI `A-testsuite` is for `src/test`, i.e. tests for the rust compiler itself, not for `#[test]`)

**Comment by SergioBenitez** [comment]
> Looks good. Thanks!
> 
> Did you look at this failure from the original PR? [#75172 (comment)](https://github.com/rust-lang/rust/pull/75172#issuecomment-675881357) Would that still be a problem?

I thought there was a chance that the `LOCAL_STREAMS` check would resolve that issue. We'll see what the CI says.

File: library/test/src/helpers/sink.rs (hunk 1 above)
Comment (SergioBenitez) [reply library/test/src/helpers/sink.rs]: Done.

**Comment by Dylan-DPC** [comment]
failed in [rollup log](https://github.com/rust-lang-ci/rust/runs/1306465020) (on wasm) 

@bors r-

File: src/test/ui/test-thread-capture.rs [hunk 2]
```diff
@@ -0,0 +1,30 @@
+// compile-flags: --test
+// run-fail
+// run-flags: --test-threads=1
+// check-run-results
+// exec-env:RUST_BACKTRACE=0
```
Comment (m-ou-se) [inline src/test/ui/test-thread-capture.rs:5]:
```suggestion
// exec-env:RUST_BACKTRACE=0
// ignore-emscripten no threads support
```

**Comment by m-ou-se** [comment]
This is now merged. Thanks for reviving this!

**Comment by SergioBenitez** [comment]
> This is now merged. Thanks for reviving this!

Thank you for the swift help!

**Pull request merged**